src/codegen/codegen-expression.lisp

```diff
@@ -107,6 +110,92 @@
+    (declare (type tc:environment env)
+             (type (or null symbol) current-function))
+
+    (let ((pred-expr (codegen-expression (node-while-expr expr) current-function env))
+          (body-expr (codegen-expression (node-while-body expr) current-function env)))
+      `(loop
+         :named ,+break-label+ 
+         :while ,pred-expr
+         :do (block ,+continue-label+ ,body-expr))))
+
+  (:method ((expr node-while-let) current-function env)
+    (declare (type tc:environment env)
+             (type (or null symbol) current-function))
+
+    (let ((match-expr (codegen-expression (node-while-let-expr expr) current-function env))
+          (body-expr (codegen-expression (node-while-let-body expr) current-function env))
+          (match-var (gensym "MATCH")))
+
+      (multiple-value-bind (pred bindings)
+          (codegen-pattern (node-while-let-pattern expr) match-var env)
+        `(loop
+           :named ,+break-label+
+           :for ,match-var := ,(if settings:*emit-type-annotations*
+                                   `(the ,(tc:lisp-type (node-type (node-while-let-expr expr)) env) ,match-expr)
+                                   match-expr)
+           :while ,pred
+           :do (block ,+continue-label+
+                 ,(cond ((null bindings) body-expr)
+                        (t `(let ,bindings
+                              (declare (ignorable ,@(mapcar #'car bindings)))
+                              ,body-expr))))))))
+
+
+  (:method ((expr node-for) current-function env)
+    (declare (type tc:environment env)
+             (type (or null symbol) current-function))
+
+    (let ((body-expr (codegen-expression (node-for-body expr) current-function env))
+          (iter-expr (codegen-expression (node-for-iter expr) current-function env))
+          (iter-var (gensym "ITERATOR"))
+          (match-var (gensym "MATCH"))
+          (next! (util:find-symbol "NEXT!" (find-package "COALTON-LIBRARY/ITERATOR"))))
+
+      (multiple-value-bind (pred bindings)
+          (codegen-pattern (node-for-pattern expr) match-var env)
+
+        `(loop
+           :named ,+break-label+
+           :with ,iter-var := ,iter-expr
+           :for ,match-var := (,next! ,iter-var)
+           :while ,pred
+           :do (block ,+continue-label+
+                 ,(cond ((null bindings) body-expr)
+                        (t `(let ,bindings
+                              (declare (ignorable ,@(mapcar #'car bindings)))
+                              ,body-expr))))))))
+  
+  (:method ((expr node-loop) current-function env)
+    (declare (type tc:environment env)
+             (type (or null symbol) current-function))
+    (let ((body-expr (codegen-expression (node-loop-body expr) current-function env))
+          (label (node-loop-label expr)))
+      `(block ,+break-label+
+         (loop ,@(when label (list :named label))
+               :do ,(if label
+                        `(tagbody ,label
```

Comment:
Is there a reason this can't be
```
`(block ,label
    (block ,+continue-label
      ,body-expr))
```

Reply:
This is to let `(continue :my-special-little-loop)` do the right thing. We use `tagbody` so that we can use `go`, otherwise we'd only hav `return-from` at our disposal and `(continue :my-special-little-loop)` would generate the same code as `(break :my-special-little-loop)`. 


Consider:

``` lisp

`(block ,+break-label+
    (loop :named :my-special-little-loop 
         :do (block ,:my-special-little-loop
                  (block ,+continue-label+
                       (return-from :my-special-little-loop) ;; is this a break or a continue ???
                        (print "not gonna happen")))))
```

If that `return-from` were generated from a `continue`, then it behaves as we expect, but if was generated from a `break` then it doesn't. That is `return-from` returns from the closest enclosing block with the name. If we want to `break` from the loop by label, then we can't; we'll always end up breaking from the inner block named `:my-special-little-loop`.

Reply:
That makes sense. I wonder if it would be easier to have all loops and all breaks/continues be labelled. I think that would reduce the number of code paths.

Reply:
I sort of compromised. 

If a loop is given a label, then we generate something like the above (but without the use of tagbody and go, instead I'm turning a label into a "continue label" or a "break label" and inserting them properly at codegen).  If a loop isn't labelled, then we generate less wrapped code using default labels. 

e.g. 

```
(loop (break))
```

generates something like 

``` 
(loop :named :coalton-loop-break 
          :do (block :coalton-loop-continue 
                      (return-from :coalton-loop-break)))

```

whereas 

``` 

(loop :x (break))

```

generates something like 

``` 

(block :coalton-loop-break 
   (loop :named :x-break 
             :do (block :coalton-loop-continue 
                        (block :x-continue 
                              (return-from :coalton-loop-break)))))
```

And the reason is I hope clear: we want to support both labelled and unlabelled breaks inside labelled loops. 

My expectation is that for most code that makes use of break and continue, unlabelled variants will be sufficient.